lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -2264,12 +2264,216 @@ struct FoldReadWritePorts : public mlir::RewritePattern {
+    DenseMap<unsigned, unsigned> mapping;
+
+    // Find which bits are used out of the users of a read port. This detects
+    // ports whose data/rdata field is used only through bit select ops. The
+    // bit selects are then used to build a bit-mask. The ops are collected.
+    SmallVector<BitsPrimOp> readOps;
+    auto findReadUsers = [&](Value port, StringRef field) {
+      auto portTy = port.getType().cast<BundleType>();
+      auto fieldIndex = portTy.getElementIndex(field);
+      assert(fieldIndex && "missing enable flag on memory port");
+
+      for (auto *op : port.getUsers()) {
+        auto portAccess = cast<SubfieldOp>(op);
+        if (fieldIndex != portAccess.getFieldIndex())
+          continue;
+
+        for (auto *user : op->getUsers()) {
+          auto bits = dyn_cast<BitsPrimOp>(user);
+          if (!bits) {
+            usedBits.set();
+            continue;
+          }
+
+          usedBits.set(bits.getLo(), bits.getHi() + 1);
+          mapping[bits.getLo()] = 0;
+          readOps.push_back(bits);
+        }
+      }
+    };
+
+    // Finds the users of write ports. This expects all the data/wdata fields
+    // of the ports to be used solely as the destination of strict connects.
+    // If a memory has ports with other uses, it is excluded from optimisation.
+    SmallVector<StrictConnectOp> writeOps;
+    auto findWriteUsers = [&](Value port, StringRef field) -> LogicalResult {
+      auto portTy = port.getType().cast<BundleType>();
+      auto fieldIndex = portTy.getElementIndex(field);
+      assert(fieldIndex && "missing enable flag on memory port");
+
+      for (auto *op : port.getUsers()) {
+        auto portAccess = cast<SubfieldOp>(op);
+        if (fieldIndex != portAccess.getFieldIndex())
+          continue;
+
+        auto conn = getSingleConnectUserOf(portAccess);
+        if (!conn)
+          return failure();
+
+        writeOps.push_back(conn);
+      }
+      return success();
+    };
+
+    // Traverse all ports and find the read and used data fields.
+    for (auto [i, port] : llvm::enumerate(mem.getResults())) {
+      // Do not simplify annotated ports.
+      if (!mem.getPortAnnotation(i).empty())
+        return failure();
+
+      // Skip debug and ports.
+      switch (mem.getPortKind(i)) {
+      case MemOp::PortKind::Debug:
+        return failure();
+      case MemOp::PortKind::Write:
+        if (failed(findWriteUsers(port, "data")))
+          return failure();
+        continue;
+      case MemOp::PortKind::Read:
+        findReadUsers(port, "data");
+        continue;
+      case MemOp::PortKind::ReadWrite:
+        if (failed(findWriteUsers(port, "wdata")))
+          return failure();
+        findReadUsers(port, "rdata");
+        continue;
+      }
+      llvm_unreachable("unknown port kind");
+    }
+
+    // Perform the transformation is there are some bits missing. Unused
+    // memories are handled in a different canonicalizer.
```

Comment:
Can we unity UnusdePorts canonicalization?

Reply:
Maybe at a later time. Now there are some discrepancies regarding the treatment of mask bits and the rewriting of ports.